---
title: "Comparison to Ruijuan's Results"
author: "Lynn Ly"
date: "2/22/2018"
output: html_document
---

# Comparisons between Lynn's ASE and DE results, and comparisons between Lynn's and Ruijuan's results
Also, comparisons between ASE results and cis / trans-eqtl results from the F2 population

```{r, include = FALSE}
library(tidyverse)
library(sleuth)
library(MBASED)
```

### Helper Functions
```{r}
ExtractGenes <- function(obj, ...) {
  UseMethod('ExtractGenes')
}

ExtractGenes.sleuth <- function(so, significant = TRUE) {
  sleuth_table <- sleuth_results(so, 'reduced:full', 'lrt', show_all = FALSE)
  if (significant == FALSE) {
    sleuth_insignificant <- dplyr::filter(sleuth_table, qval > 0.05)
    return(sleuth_insignificant$target_id)
  } else {
    sleuth_significant <- dplyr::filter(sleuth_table, qval <= 0.05)
    return(sleuth_significant$target_id)
  }
}

ExtractGenes.data.frame <- function(df) {
  if("genes" %in% colnames(df)) {
    return(df$genes)
  } else if ("target_id" %in% colnames(df)) {
    return(df$target_id) 
  } else if ("GeneID" %in% colnames(df)) {
    return(df$GeneID)
  } else if ("trait" %in% colnames(df)) {
    return(df$trait)
  } else {
    return(rownames(df))
  }
}

CommonGenes <- function(list1, list2) {
  name1 <- deparse(substitute(list1))
  name2 <- deparse(substitute(list2))
  matches <- intersect(list1, list2)
  cat("\n", name1, " has ", length(list1), " genes.")
  cat("\n", name2, " has ", length(list2), " genes.")
  cat("\n", length(matches), " genes in common.")
  return(matches)
}

```


## Differential Expression Lynn's vs Ruijuan's

Comparison between Lynn's sleuth and Ruijuan's DE genes from edgeR
```{r}
so <- sleuth_load("sleuthResults.Rdata")
load("DEgene.gt.F1.Rdata") # Ruijuan's edgeR results

CommonGenes(ExtractGenes(so), ExtractGenes(DEgene.young.F1))
```

We used a different filtering method for the edgeR analysis, keeping only genes with at least 10 reads in at least 3 samples.  
Since I only looked at DE in young samples, I need to use the old read counts to get the same filtered reads.  
```{r}
reads <- read.table("read.count.tsv", header = T, check.names = F)

# filter based on read count 
filteredReads <- reads[rowSums(reads[,-1] > 10) >= 3,]
dim(filteredReads) #60176 genes remain

sleuth_significant_filtered <- semi_join(sleuth_significant, filteredReads, by = c("target_id" = "ID"))
sleuth_significant_removed <- anti_join(sleuth_significant, filteredReads, by = c("target_id" = "ID"))
dim(sleuth_significant_filtered) # 2316 genes passed both filters
dim(sleuth_significant_removed) # 54 genes passed sleuth's basic filter, but not our edgeR filter.
```

Comparison between Lynn's edgeR on SNV-only subset and Ruijuan's DE genes from edgeR (subsetted after running)
```{r}
load("youngDEGenes_subset.Rdata") # Lynn's edgeR genes, subsetted before running
load("phasedData.Rdata") # For getting SNV-inclusive genes, to subset Ruijuan's edgeR results post-running

dim(DEgene.young.F1) #4024 genes 
dim(youngDEGenes_subset) # 1432 genes

DEgene.young.F1_subset <- semi_join(DEgene.young.F1, phasedData, by = c("target_id" = "GeneID"))
dim(DEgene.young.F1_subset) # 986 genes

matches <- semi_join(DEgene.young.F1_subset, youngDEGenes_subset, by = c("target_id" = "genes"))
dim(matches) 
# All 986 genes that were significant without subsetting before running remained significant after subsetting.
# In addition, 446 more genes were found to be in significant DE
```

## Allele Specific Expression Lynn's vs Ruijuan's

Loading Ruijuan's results using a different method, which used a binomial model only

```{r}
load("ASE_gene.Rdata") # Ruijuan's results
length(ASE_gene) # 2530 genes found in ASE by Ruijuan

load("aseGenesBinom") # 666 genes
load("aseGenesBeta") # 96 genes

matches <- intersect(ASE_gene, aseGenesBinom)
length(matches) # 554 out of 666 genes agreed using binomial model

matches <- intersect(ASE_gene, aseGenesBeta)
length(matches) # 70 out of 96 genes using beta-binomial model
```

## Lynn's ASE and DE Results

### Comparison to allele specific expression results in MBASED

Comparison to ASE results when **ONLY considering genes that contain a SNV.**  
sleuth was run after filtering for genes with SNVs.  
no enrichment in my data
```{r}
so <- sleuth_load("sleuthResultsSubset.Rdata")

DEgenes <- ExtractGenes(so, significant = TRUE)
nonDEgenes <- ExtractGenes(so, significant = FALSE)

load("aseGenesBeta.Rdata")

diffASEyesDE <- intersect(aseGenesBeta, DEgenes)
sameASEyesDE <- setdiff(DEgenes, aseGenesBeta)
diffASEnoDE <- setdiff(aseGenesBeta, DEgenes)
sameASEnoDE <- setdiff(nonDEgenes, aseGenesBeta)

cat(length(diffASEyesDE), "\t", length(sameASEyesDE), "\n", length(diffASEnoDE), "\t" ,length(sameASEnoDE))

ASE_DE_SNP <- data.frame(DE = c(10, 912), 
                         no_DE = c(86, 10385))

rownames(ASE_DE_SNP) = c("ASE", "no_ASE")

fisher.test(ASE_DE_SNP) 
```

Same test, using Lynn's subsetted edgeR results instead of sleuth
```{r}
load("youngDEGenes_subset.Rdata") # Lynn's edgeR genes, subsetted before running
load("phasedData.Rdata") # For getting SNV-inclusive genes

allGenes <- ExtractGenes(phasedData)
DEgenes <- ExtractGenes(youngDEGenes_subset)
nonDEgenes <- setdiff(allGenes, DEgenes)

diffASEyesDE <- intersect(aseGenesBeta, DEgenes)
sameASEyesDE <- setdiff(DEgenes, aseGenesBeta)
diffASEnoDE <- setdiff(aseGenesBeta, DEgenes)
sameASEnoDE <- setdiff(nonDEgenes, aseGenesBeta)

cat(length(diffASEyesDE), "\t", length(sameASEyesDE), "\n", length(diffASEnoDE), "\t" ,length(sameASEnoDE))

ASE_DE_SNP <- data.frame(DE = c(16, 1416), 
                         no_DE = c(80, 10121))

rownames(ASE_DE_SNP) = c("ASE", "no_ASE")

fisher.test(ASE_DE_SNP)
```


## Comparisons to eQTL lists

Modified version of AnnotatedSNPs, just to get a list of genes that could be possible eQTLs
```{r}
AnnotateeQTLs <- function(eQTLdata, gff.mRNA = gff.mRNA){
  # Takes the list of eQTLs, and appends genes that fall within the eQTL interval. 
  eQTLdata <- eQTLdata[, c(1, 3, 5, 6)]
  colnames(eQTLdata) <- c("trait", "CHROM", "START", "END")
  colnames(gff.mRNA) <- c("CHROM", "start", "end", "name") 
  
  genes <- GRanges(seqnames = Rle(gff.mRNA$CHROM),
                   ranges = IRanges(start = gff.mRNA$start, end = gff.mRNA$end), 
                   names = gff.mRNA$name)
  
  SNPs <- GRanges(seqnames = Rle(eQTLdata$CHROM), 
                 ranges = IRanges(start = eQTLdata$START, eQTLdata$END), 
                 CHROM = eQTLdata$CHROM,
                 START = eQTLdata$START,
                 END = eQTLdata$END)
  
  # Overlap SNP position with gene range 
  overlappedGenes <- mergeByOverlaps(SNPs, genes)
  colnames(overlappedGenes) <- c("eQTLRanges", "CHROM", "START", "END", "GeneRanges", "Genes")
  overlappedGenes <- overlappedGenes[, c(2, 3, 4, 6)]
  
  annotatedeQTLdata <- eQTLdata %>% 
    left_join(as.data.frame(overlappedGenes), by=c("CHROM", "START", "END")) 
  
  cat("\n", length(unique(annotatedeQTLdata$Genes)), "genes overlap with the potential eQTL intervals")
  
  return(annotatedeQTLdata)  
  
}

FiltereQTLs <- function(eQTLs, genelist) {
  # Takes the list of eQTLs, and returns the list subsetted to contain only genes in genelist that have an eQTL.
  eQTLgenes <- ExtractGenes(eQTLs)
  cat("\n", length(intersect(eQTLgenes, genelist)), " out of ", length(genelist)," genes had an eQTL") 
  eQTLs <- eQTLs[which(eQTLs$gene_ID %in% genelist), ]
  return(eQTLs)
}


```

Ruijuan asks:
1) Are the genes which show DE being trans-regulated? to do this, check the overlap of genes in the trans-eQTL list with the DEGs list excluding those showing different ASE.

2) If yes, many of them are overlapped, then what are the genes/loci that regulate these trans-regulated DEGs? to do this, check the eQTL interval for these trans-regulated DE genes, among those eQTL interval, do we found genes exhibit differential ASE? are those genes being cis-regulated? (i.e, are they included in the cis-eQTL gene list?)

```{r}
load("cis_trans_result_new_flipped.Rdata") # eQTL lists from F2
gff.mRNA <- read.table("/Network/Servers/avalanche.plb.ucdavis.edu/Volumes/Mammoth/Users/ruijuanli/Reference/B.napus/gff.mRNA")

eQTLgenes <- trans_eQTL$gene_ID
ciseQTLgenes <- cis_eQTL$gene_ID

trans_eQTL_DE <- intersect(eQTLgenes, DEgenes) #360 out of 1432 DEGs are regulated by trans-eQTLs
cis_eQTL_DE <- intersect(ciseQTLgenes, DEgenes) # 758 are regulated by cis-eQTLs
cis_trans_DE <- intersect(cis_eQTL_DE, trans_eQTL_DE) # 128 in both lists

trans_eQTL_ASE <- intersect(eQTLgenes, aseGenesBeta)
cis_eQTL_ASE <- intersect(ciseQTLgenes, aseGenesBeta)
cis_trans_ASE <- intersect(cis_eQTL_ASE, trans_eQTL_ASE)
```

Filter to only examine trans-eQTL intervals for these 360 DEGs
THIS IS NOT USEFUL. Almost all genes are represented in the trans-eQTL intervals.
```{r}
trans_eQTL_DE <- AnnotateeQTLs(FiltereQTLs(trans_eQTL, DEgenes), gff.mRNA)
cis_eQTL_DE <- AnnotateeQTLs(FiltereQTLs(cis_eQTL, DEgenes), gff.mRNA)

trans_eQTL_ASE <- AnnotateeQTLs(FiltereQTLs(trans_eQTL, aseGenesBeta), gff.mRNA)
cis_eQTL_ASE <- AnnotateeQTLs(FiltereQTLs(cis_eQTL, aseGenesBeta), gff.mRNA)
```




