---
title: "Sleuth Differential Expression Analysis"
author: "Lynn Ly"
date: "February 16, 2018"
output: html_document
---

```{r setup, include=FALSE}
library(sleuth)
```

## Preliminaries

```{r}
sample_id <- dir(file.path("results"))

kal_dirs <- file.path("results", sample_id, "kallisto")
kal_dirs
```

Get the metadata from our experiment summary table and add a new column, path, leading to the kallisto directories. 
```{r}
metadata <- read.csv(file.path("F1_summary.csv"), header = TRUE, stringsAsFactors=FALSE)
metadata <- metadata[1:6, ]# young tissue only
metadata <- dplyr::mutate(metadata, path = kal_dirs)
metadata <- dplyr::select(metadata[1:6, ], sample = Sample.ID, cultivar, path, rep)
```


Include gene-names into transcript-level analysis

**Not necessary, as we included the gene names when we created the kallisto index. However, we cannot do transcript-level analysis, only gene leve**
```{r}
# bnapus <- biomaRt::useMart(biomart = "plants_mart", dataset = "bnapus_eg_gene", host = "plants.ensembl.org")
# listAttributes(bnapus)
# t2g <- biomaRt::getBM(attributes = c("ensembl_transcript_id", "ensembl_gene_id", "external_gene_name"), mart = bnapus)
# t2g <- dplyr::rename(t2g, target_id = ensembl_transcript_id, ens_gene = ensembl_gene_id, ext_gene = external_gene_name)
```


Initialize sleuth object

```{r cache=TRUE}
so <- sleuth_prep(metadata, extra_bootstrap_summary = TRUE)
```

Full model 

```{r cache=TRUE}
so <- sleuth_fit(so, ~cultivar, 'full')
so <- sleuth_fit(so, ~rep + cultivar, 'fullRep')
```
Reduced model

```{r cache=TRUE}
so <- sleuth_fit(so, ~1, 'reduced')
so <- sleuth_fit(so, ~rep, 'reducedRep')
```

Compare the two models

```{r}
so <- sleuth_lrt(so, 'reduced', 'full')
so <- sleuth_lrt(so, 'reducedRep', 'fullRep')

#save(so, file = "SleuthOutput.Rdata")
#save(metadata, file = "metadata.Rdata")
load("SleuthOutput.Rdata")
#load("metadata.Rdata")
```

In general, sleuth can utilize the likelihood ratio test with any pair of models that are nested
The results of the test can be examined with
```{r}
sleuth_table <- sleuth_results(so, 'reduced:full', 'lrt', show_all = FALSE)
sleuth_significant <- dplyr::filter(sleuth_table, qval <= 0.05)

sleuth_table_rep <- sleuth_results(so, 'reducedRep:fullRep', 'lrt', show_all = FALSE)
sleuth_significant_rep <- dplyr::filter(sleuth_table_rep, qval <= 0.05)


difference <- anti_join(sleuth_significant_rep, sleuth_significant, by = "target_id")
same <- semi_join(sleuth_significant_rep, sleuth_significant, by = "target_id")

head(sleuth_significant, 20)
dim(sleuth_significant) # Genes that are significantly differentially expressed
```

The table shown above displays the top 20 significant genes with a (Benjamini-Hochberg multiple testing corrected) q-value <= 0.05.

```{r}
plot_bootstrap(so, "BnaC03g34410D", units = "est_counts", color_by = "cultivar")
```


The easiest way to view and interact with the results is to generate the sleuth live site that allows for exploratory data analysis:

```{r eval=FALSE}
sleuth_live(so)
```

Among the tables and visualizations that can be explored with sleuth live are a number of plots that provide an overview of the experiment. For example, a PCA plot provides a visualization of the samples:
```{r}
plot_pca(so, color_by = 'cultivar')
```

Various quality control metrics can also be examined. The count distributions for each sample (grouped by condition) can be displayed using the `plot_group_density` command:

```{r}
plot_group_density(so, use_filtered = TRUE, units = "est_counts",
  trans = "log", grouping = setdiff(colnames(so$sample_to_covariates),
  "sample"), offset = 1)
```


