---
title: "F2 SNPs"
author: "Lynn Ly"
date: "3/16/2018"
output: html_document
---

Inputs: A VCF file (input/F2_filtered/vcf) and filtering criteria (input/F2_Final_SNP_Calls, ../MBASED/aseGenes414)
Outputs: A tidy VCF file, filtered. Includes heterozygous SNVs associated with genes in ASE. (tidy_vcf/aseRatiosF2)

```{r}
library(ggplot2)
library(vcfR)
```

### Function for extracting relevant elements from a tidy vcf
```{r}
tidyVCF2ASE <- function(tidyVCF, finalCalls) {
  fix <- tidyVCF$fix[, c("ChromKey", "CHROM", "POS", "REF", "ALT")] 
  gt <- tidyVCF$gt[, c("ChromKey", "POS", "Indiv", "gt_GT", "gt_DP", "gt_RO", "gt_AO")]
  factors <- c("Indiv", "gt_GT")
  gt[factors] <- lapply(gt[factors], factor)
  gt$gt_AO <- as.integer(gt$gt_AO)
  
  # Filtering for specific SNVs
  finalCalls <- left_join(finalCalls[, c(1, 2, 5, 6)], tidyVCF$fix[, 1:3], by = c("CHROM", "POS"))
  fix <- inner_join(fix, finalCalls, by = c("ChromKey", "POS", "CHROM"))
  gt <- semi_join(gt, finalCalls, by = c("ChromKey", "POS"))
  counts <- inner_join(fix, gt, by = c("ChromKey", "POS"))
  
  ASE <- counts %>% mutate(
    gt_RR = gt_RO / gt_DP,
    gt_AR = gt_AO / gt_DP
  )
  
  colnames(ASE)[6:13] <- c("Ae", "Ol", "Indiv", "GT", "DP", "RO", "AO", "RR", "AR")
  
  return(ASE)
}
```

### Convert our VCF file into a tidy df for downstream analysis  
The input VCF contained extra loci. It was filtered to only include loci in F2_Final_SNP_Calls.  
```{r}
vcf <- read.vcfR("input/F2_filtered.vcf")  

info_fields <- c("DP")
format_fields <- c("GT", "DP", "RO", "AO")
tidyVCF <- vcfR2tidy(vcf, info_fields = info_fields, format_fields = format_fields)

save(tidyVCF, file = "tidy_vcf/tidyVCF.Rdata")

finalCalls <- read.table("input/F2_Final_SNP_Calls", header = TRUE)

aseRatiosF2_full <- tidyVCF2ASE(tidyVCF, finalCalls)
save(aseRatiosF2_full, file = "tidy_vcf/aseRatiosF2_full.Rdata")
```

## Filtering 
Only include SNVs where the individual is heterozygous.
```{r}
load("tidy_vcf/aseRatiosF2_full.Rdata")

summary(aseRatiosF2_full$GT) # 1246299 are heterozygous

aseRatiosF2 <- aseRatiosF2_full %>% filter(GT == "0/1")
```

Also, only include SNVs from genes previously determined to be in ase, from the F1414 population.
```{r}
load("../MBASED/aseGenes414.Rdata") # Genes determined to be in ASE
load("../MBASED/phasedData.Rdata")

phasedData_filtered <- phasedData %>% filter(GeneID %in% aseGenes414)
aseRatiosF2 <- aseRatiosF2 %>% semi_join(phasedData_filtered, by = c("CHROM", "POS"))

save(aseRatiosF2, file = "tidy_vcf/aseRatiosF2.Rdata")
```

