---
title: "edgeR_F1.rmd"
author: "Lynn Ly"
date: "2/22/2018"
output: html_document
---

```{r setup, include=FALSE}
library(edgeR)
library(tidyverse)
```

Set up 
```{r}
reads <- read.table("read.count.tsv", header = TRUE)
rownames(reads) <- reads$ID
reads <- reads[, -1]

metadata <- read.csv("F1_summary.csv")
metadata <- metadata[order(metadata$Sample.ID),]

metadata[,5:8]
sampleIDs <- paste(metadata$cultivar, metadata$Stage, metadata$rep, sep = "_")
sampleIDs

colnames(reads) <- sampleIDs
head(reads)
save(reads, file="differential_expression/edgeRReads.Rdata")
```

### Filtering

In this edgeR rerun, we only want to include genes that have SNVs in them.
```{r}
load("differential_expression/edgeRReads.Rdata")
load("MBASED/phasedData.Rdata") # This is the annotated data from MBASED analysis. It has already gone through SNP calling 

geneIDs <- as.character(unique(phasedData$GeneID))

# Only genes with at least 10 reads in at least 3 samples
reads_subset <- reads[rowSums(reads > 10) >= 3,]
dim(reads_subset) # 60176  24 

keep <- rownames(reads_subset) %in% geneIDs
reads_subset <- reads_subset[keep, ] # 48895 genes discarded. 11281 genes remaining
save(reads_subset, file = "differential_expression/edgeRReadsSubset.Rdata")
```

Creating and normalizing dgList
```{r}
load("differential_expression/edgeRReadsSubset.Rdata")

metadata$group <- paste(metadata$cultivar, metadata$Stage, sep = "_")

dgList <- DGEList(counts = reads_subset, genes = rownames(reads_subset), group = metadata$group)

# Normalize
dgList <- calcNormFactors(dgList, method = "TMM") 

group <- metadata$group
designMatrix <- model.matrix(~0 + group, data = reads_subset)

# Check that clustering makes sense, no outliers
plotMDS(dgList)
```


Estimating Dispersions
```{r}
# Estimate the dispersion parameter to be used in our negative binomial model
dgList <- estimateDisp(dgList, design = designMatrix)

plotBCV(dgList)
```


Differential Expression between young tissue
```{r}
fit <- glmFit(dgList, designMatrix)
colnames(fit$design)

# Comparing between 414 and 415 young tissue, using 414 as reference
youngLRT <- glmLRT(fit, contrast = c(0,0,0,-1,0,0,0,1))
topTags(youngLRT) 

youngDE <- decideTestsDGE(youngLRT, p = 0.05)
summary(youngDE) # 668 downreg, 9849 neither, 764 upreg

# A sorted list of all the significant DE genes
edgeRResultsYoung_subset <- topTags(youngLRT,n = Inf)$table[topTags(youngLRT,n = Inf)$table$FDR<0.05,]

save(edgeRResultsYoung_subset, file = "differential_expression/edgeRResultsYoung_subset.Rdata")
```


